format-common/src/main/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingDecompressor.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.charset.fixedlength;
+
+import org.apache.hadoop.io.compress.Decompressor;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Decompressor that can be used to convert byte streams in fixed-length character encodings to a stream of UTF-8 bytes.
+ */
+public class FixedLengthCharsetTransformingDecompressor implements Decompressor {
+
+  final FixedLengthCharset origin;
+  final Charset destination = StandardCharsets.UTF_8;
+  long numConsumedBytes = 0;
+
+  ByteArrayOutputStream incomingBuffer = new ByteArrayOutputStream();
+  ByteArrayInputStream outgoingBuffer = new ByteArrayInputStream(new byte[]{});
+
+  public FixedLengthCharsetTransformingDecompressor(FixedLengthCharset origin) {
+    this.origin = origin;
+  }
+
+  @Override
+  public void setInput(byte[] b, int off, int len) {
```

Comment:
You are doing a lot of copying here. Please consider using java.nio.charset.Charset along with `ByteBuffer`/`CharBuffer`.

Reply:
Hmm, good point. I can probably get away with using a byte buffer for input and transform as needed when decompression is needed.